cranelift/codegen/src/isa/s390x/lower.isle

```diff
@@ -0,0 +1,1766 @@
+
+;; Implement `sdiv`.
+(rule (lower (has_type (fits_in_64 ty) (sdiv x y)))
+      (let (;; Look at the divisor to determine whether we need to generate
+            ;; explicit division-by-zero and/or integer-overflow checks.
+            (DZcheck bool (zero_divisor_check_needed ty y))
+            (OFcheck bool (div_overflow_check_needed y))
+            ;; Load up the dividend (sign-extended to 64-bit) into the low
+            ;; half of a register pair (the high half remains uninitialized).
+            (ext_x RegPair (put_in_regpair_lo_sext64 x (uninitialized_regpair)))
+            ;; Load up the divisor (sign-extended if necessary).
+            (ext_y Reg (put_in_reg_sext32 y))
+            ;; Perform division-by-zero check (same as for `udiv`).
+            (_1 Reg (maybe_trap_if_zero_divisor DZcheck ty ext_y))
+            ;; Perform integer-overflow check if necessary.
+            (_2 Reg (maybe_trap_if_sdiv_overflow OFcheck ty ext_x ext_y))
+            ;; Emit the actual divide instruction.
+            (pair RegPair (sdivmod ty ext_x ext_y)))
+        ;; The quotient can be found in the low half of the result.
+        (value_reg (copy_reg ty (regpair_lo pair)))))
+
+;; Implement `srem`.  Same as `sdiv`, but finds the remainder in
+;; the high half of the result register pair instead.  Also, handle
+;; the integer overflow case differently, see below.
+(rule (lower (has_type (fits_in_64 ty) (srem x y)))
+      (let ((DZcheck bool (zero_divisor_check_needed ty y))
+            (OFcheck bool (div_overflow_check_needed y))
+            (ext_x RegPair (put_in_regpair_lo_sext64 x (uninitialized_regpair)))
+            (ext_y Reg (put_in_reg_sext32 y))
+            (_ Reg (maybe_trap_if_zero_divisor DZcheck ty ext_y))
+            (checked_x RegPair (maybe_avoid_srem_overflow OFcheck ty ext_x ext_y))
+            (pair RegPair (sdivmod ty checked_x ext_y)))
+        (value_reg (copy_reg ty (regpair_hi pair)))))
+
+;; Determine whether we need to perform an integer-overflow check.
+;;
+;; We never rely on the divide instruction itself to trap; while that trap
+;; would indeed happen, we have no way of signalling two different trap
+;; conditions from the same instruction.  By explicity checking for the
+;; integer-overflow case ahead of time, any hardware trap in the divide
+;; instruction is guaranteed to indicate divison-by-zero.
+;;
+;; In addition, for types smaller than 64 bits we would have to perform
+;; the check explicitly anyway, since the instruction provides a 64-bit
+;; quotient and only traps if *that* overflows.
+;;
+;; However, the only case where integer overflow can occur is if the
+;; minimum (signed) integer value is divided by -1, so if the divisor
+;; is any immediate different from -1, the check can be omitted.
+(decl div_overflow_check_needed (Value) bool)
+(rule (div_overflow_check_needed (i64_from_value -1)) $true)
+(rule (div_overflow_check_needed (i64_from_value _)) $false)
```

Comment:
A future-looking note here: when we get to verifying rules, in theory we want every rule to be independently correct, i.e., not only correct in cases not shadowed by other rules. Here and in some other spots nearby I notice the pattern of specific cases followed by wildcards, which is fine for this as long as the more general fallback is correct without the special case, but here appears not to be.

Earlier we had tried to follow the principle that rule ordering is *completely* flexible, and any applicable rule could fire, so the LHSes need to be disjoint or priorities need to be used if one must fire over the other. But I think that's pretty surprising and un-ergonomic in cases like this, so the DSL compiler now goes for "more specific rule first" which will indeed to the right thing here.

We'll undoubtedly uncover other cases like this if/when we do try to verify in this way but I just thought I'd note it as a principle we're trying to stick to elsewhere; here a `(not_neg1)` extractor or somesuch might cause fewer issues later...

(Part of me still likes the simplicity of "first written rule takes priority"; it's simple, explicit, no hidden magic... but others voiced good reasons not to do that, so, too late now probably...)

Reply:
> We'll undoubtedly uncover other cases like this if/when we do try to verify in this way but I just thought I'd note it as a principle we're trying to stick to elsewhere; here a (not_neg1) extractor or somesuch might cause fewer issues later...

+1

Reply:
OK, I wasn't aware of that requirement.  For the divide checks, this should be straightforward using some extra extractors as you say.   We still need a default case, but as long as the default is `true` instead of `false` it will still be always correct.  However, I guess I then cannot use the `(avoid_div_traps)` extractor as-is, since I need the negated sense now.  I don't think there's way to do that negation on ISLE, so I'd need another extractor e.g. like so:
```
(decl zero_divisor_check_needed (Value) bool)
(rule (zero_divisor_check_needed (u64_from_value (u64_nonzero _)) $false))
(rule (zero_divisor_check_needed (value_type (allow_div_traps)) $false))
(rule (zero_divisor_check_needed _ $true))
```
That new `(allow_div_traps)` should probably go into the common prelude as well.

However, I have made that same assumption extensively elsewhere, in particular when dispatching between different input types.  For example, there's frequently the pattern of a pair of  `(fits_in_32 ty)` and `(fits_in_64 ty)` rules, where 32-bit types *must* use the `fits_in_32` rule, since the input isn't sufficiently sign- or zero-extended so the instruction behind the `fits_in_64` rule would yield an incorrect result.   Similarly, for e.g. moves I have both a `(fits_in_32 ty)` rule and a `$F32` rule, where the `fits_in_32` rule is incorrect for `$F32` as it uses a GPR instruction.

To ensure each rule is independently correct (if not optimal), I need to rework all this type handling.  This probably means making supported types for each rule much more explicit, which is probably a good thing in the long run anyway.

I'll be working on this and provide an updated patch.